### DISCUSSION TITLE
defines and includePath in c_cpp_properties.json

### DISCUSSION THREAD

**Discussion by mkitby** [microsoft/vscode-cpptools#7579] [Q&A]
Hi,

defines and includePath settings will be discarded if using compile_commands.json in c_cpp_properties.json. I think this is unreasonable. Both settings should be kept. Let's think about a cross compiler which is not supported by cpptools. In this case, the compilerPath setting will be defined as an empty string to skip querying a compiler. And compiler internal defines will NOT exist in compile_commands.json. We need to define it in defines setting.

**Comment by sean-mcmanus** [comment]
Cross compilation is supported in general...what issue are you hitting with that? We appear to be tracking your request with https://github.com/microsoft/vscode-cpptools/issues/6909 .

**Reply by mkitby** [reply]
same issue as #6909

**Comment by lakazatong** [comment]
I would like to add a follow up on that topic, compiling a project can be troublesome, whereas adding WIP header files in the includePath array is much easier and quicker.
Moreover as @mkitby said, it'"unreasonable" to discard include Path if a compile_commands.json is used, I'd even say it's unexpected behavior, I would expect both to complement each other, not one to override the other.
Imagine all of your includes working all of a sudden because you deleted your build directory... what a counter-intuitive side effect.

**Comment by DakotaB75** [comment]
This is expected behavior when `compile_commands.json` is present.
cpptools intentionally ignores `includePath`, `defines` and `compilerPath` in that case, because the compilation database becomes the single source of truth for IntelliSense. This avoids mismatches between IntelliSense and the actual compiler invocation.
Where this feels counter-intuitive (especially for cross-compilation or WIP headers), the correct approach is to **extend the compilation database itself**, not mix configuration layers.

Example:

```json
"command": "arm-none-eabi-gcc -I./wip_headers -DDEBUG ..."
```
or generate it properly from the build system (CMake, Meson, etc.).
An additive/opt-in override mode would improve DX, but until then this behavior is by design (tracked in #6909).

**Reply by sean-mcmanus** [reply]
@DakotaB75 FYI, there is a mergeConfigurations that works with a configuration provider, but it hasn't been implemented to work for compile_commands.json yet.

**Comment by DakotaB75** [comment]
Thanks for the clarification that aligns with the current behavior. The confusion mainly comes from compile_commands.json being treated differently than other configuration providers. Having an opt-in additive mode (or clearer docs) would likely prevent this class of issues for cross-compile and WIP headers.
